nemo/collections/speechlm2/models/duplex_ear_tts.py

```diff
@@ -0,0 +1,1522 @@
+
+
+def rescale_state_dict(state_dict, target_std=0.02, first_n_layers=None, layer_prefix="tts_model.backbone.layers."):
+    """
+    Rescale trainable weights in a state_dict for BF16 stability.
+
+    Args:
+        state_dict: PyTorch state_dict
+        target_std: desired target std for weights
+        first_n_layers: if not None, rescale only the first N transformer blocks
+        layer_prefix: prefix for layer names (default: "tts_model.backbone.layers.")
+    Returns:
+        new_state_dict
+    """
+    weight_tensors = []
+
+    # Compute which prefixes to match if first_n_layers is set
+    prefixes_to_match = []
+    if first_n_layers is not None:
+        prefixes_to_match = [f"{layer_prefix}{i}" for i in range(first_n_layers)]
+
+    for name, param in state_dict.items():
+        if not torch.is_tensor(param):
+            continue
+
+        if "rvq_embs" in name:
+            continue
+
+        # Skip biases & 1-dim params (norm weights/gates)
+        if param.ndim <= 1:
+            continue
+
+        # Skip layers not in the first N
+        if first_n_layers is not None and not any(name.startswith(pfx) for pfx in prefixes_to_match):
+            continue
+
+        weight_tensors.append(param.float())
+
+    if not weight_tensors:
+        if first_n_layers is not None:
+            print(f"⚠️ No weights found for first {first_n_layers} layers with prefix '{layer_prefix}'.")
+        else:
+            print("⚠️ No weights found to rescale in state_dict.")
+        return state_dict
+
+    # Compute global std across selected weights (on CPU)
+    cpu_weights = [p.detach().cpu() for p in weight_tensors]
+    flat = torch.cat([p.flatten() for p in cpu_weights])
+    current_std = float(torch.std(flat))
+    scale = target_std / (current_std + 1e-8)
+
+    print(
+        f"📦 Rescaling state_dict "
+        f"{'(first N layers)' if first_n_layers else '(all layers)'}: "
+        f"current std = {current_std:.6f}, target = {target_std}, scale = {scale:.6f}"
+    )
+
+    # Apply scaling
+    new_state_dict = {}
+    for name, param in state_dict.items():
+        if (
+            torch.is_tensor(param)
+            and param.ndim > 1
+            and (first_n_layers is None or any(name.startswith(pfx) for pfx in prefixes_to_match))
+        ):
+            new_state_dict[name] = param * scale
+        else:
+            new_state_dict[name] = param
+
+    print("✅ Done: weights rescaled.")
+    return new_state_dict
+
+
+class DuplexEARTTS(LightningModule, HFHubMixin):
+    def __init__(self, cfg: dict) -> None:
+        assert isinstance(cfg, dict), (
+            "You must pass the config to DuplexEARTTS as a Python dict to support hyperparameter serialization "
+            f"in PTL checkpoints (we got: '{type(cfg)=}')."
+        )
+        super().__init__()
+        self.save_hyperparameters()
+        # convert dict to config
+        cfg = DictConfig(cfg)
```

Comment:
Why do we pass a dict and forcibly convert it to a DictConfig rather than passing a DictConfig?

Reply:
PyTorch Lightning `save_hyperparameters` won't work with `DictConfig`; it must be `dict`. Unfortunate hack.